api/v1alpha1/addon_types.go

```diff
@@ -257,4 +257,27 @@ type ServiceLoadBalancer struct {
 	// +kubebuilder:validation:Enum=MetalLB
 	// +kubebuilder:validation:Required
 	Provider string `json:"provider"`
+
+	// Configuration for the chosen ServiceLoadBalancer provider.
+	// +kubebuilder:validation:Optional
+	Configuration *ServiceLoadBalancerConfiguration `json:"configuration,omitempty"`
+}
+
+type ServiceLoadBalancerConfiguration struct {
+	// AddressRanges is a list of IPv4 address ranges the
+	// provider uses to choose an address for a load balancer.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	AddressRanges []AddressRange `json:"addressRanges"`
+}
+
+// AddressRange defines an IPv4 range.
+type AddressRange struct {
+	// Source: https://github.com/metal3-io/ip-address-manager/blob/b4f7c2a/api/v1alpha1/common_types.go#L19
+
+	// +kubebuilder:validation:Pattern="^((([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5]))$"
```

Comment:
```suggestion
	// +kubebuilder:validation:Format=ipv4
```

Reply:
Ok. Can you please link to where this is documented?

Reply:
I appreciate the effort, but I'm looking for something authoritative.

The swagger docs says:

> However, format is an open value, so you can use any formats, even not those defined by the OpenAPI Specification

And the kubebuilder docs don't mention ipv4.

And I don't think the apiextensions code is relevant here.

It looks like the API server imports https://github.com/kubernetes/kube-openapi, but that doesn't have any documentation...

Reply:
I want to an authoritative source, so that future me (and others) can discover what formats are available.

Reply:
And please add a test for this to prove the validation works. We have examples of this in various places.

Reply:
I documented about using formats in our [API conventions](https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/blob/main/docs/content/contributing/api_conventions.md?plain=1#L55-L61) btw, including highlighting that those linked formats work.

Reply:
We already use `uri` in https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/blob/main/api/v1alpha1/clusterconfig_types.go#L244 and have tests for this in https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/blob/main/pkg/handlers/generic/mutation/mirrors/variables_test.go

Reply:
Upstream issue: https://github.com/kubernetes-sigs/kubebuilder/issues/3479